miner/block_builder.go

```diff
@@ -0,0 +1,192 @@
+	Origin			common.Address //todo: remove this, should be calculated from sig.
+	Amount       	[]byte
+	Payload      	[]byte
+}
+
+
+func NewBlockBuilder(net p2p.Service, beginRoundEvent chan mesh.LayerID, weakCoin WeakCoinProvider, orph OrphanBlockProvider) BlockBuilder{
+	return BlockBuilder{
+		beginRoundEvent:beginRoundEvent,
+		stopChan:make(chan struct{}),
+		newTrans:make(chan *state.Transaction),
+		hareResult:make(chan HareResult),
+		transactionQueue:make([]SerializableTransaction,0,10),
+		hareRes:nil,
+		mu:sync.Mutex{},
+		network:net,
+		weakCoinToss:weakCoin,
+		orphans:orph,
+		started: false,
+	}
+}
+
+func Transaction2SerializableTransaction(tx *state.Transaction) SerializableTransaction{
+	return SerializableTransaction{
+		AccountNonce:tx.AccountNonce,
+		Origin:tx.Origin,
+		Recipient:tx.Recipient,
+		Amount:tx.Amount.Bytes(),
+		Payload:tx.Payload,
+		GasLimit:tx.GasLimit,
+		Price:tx.Price.Bytes(),
+	}
+}
+
+func (t *BlockBuilder) Start() error{
+	if t.started {
+		return fmt.Errorf("already started")
+	}
+
+	t.started = true
+	go t.acceptBlockData()
+	return nil
+}
+
+func (t *BlockBuilder) Stop() error{
+	if !t.started {
+		return fmt.Errorf("already stopped")
+	}
+	t.stopChan <- struct{}{}
+	return nil
+}
+
+type HareResult interface {
+	GetResult() []mesh.BlockID
+}
+
+type WeakCoinProvider interface {
+	GetResult() bool
+}
+
+type OrphanBlockProvider interface {
+	GetOrphans() []mesh.BlockID
+}
+
+func (t *BlockBuilder) CommitHareResult(res HareResult) error{
+	if !t.started{
+		return fmt.Errorf("BlockBuilderStopped")
+	}
+	t.hareResult <- res
+	return nil
+}
+
+func (t *BlockBuilder) AddTransaction(nonce uint64, origin, destination common.Address, amount *big.Int) error{
+	if !t.started{
+		return fmt.Errorf("BlockBuilderStopped")
+	}
+	tx := state.NewTransaction(nonce,origin,destination,amount)
+	t.newTrans <- tx
+	return nil
+}
+
+func (t *BlockBuilder) resetAfterCreateBlock(){
+	t.hareRes = nil
+}
+
+func (t *BlockBuilder) createBlock(id mesh.LayerID, txs []SerializableTransaction) Block {
+	b := Block{
+		Id : mesh.BlockID(rand.Int63()),
+		LayerIndex:id,
+		Data:nil,
+		Coin: t.weakCoinToss.GetResult(),
+		Timestamp:time.Now(),
+		Txs : make([]SerializableTransaction, len(txs)),
+		HareResult : t.hareRes.GetResult(),
+		Orphans:t.orphans.GetOrphans(),
+	}
+	copy(b.Txs, txs)
+	t.resetAfterCreateBlock()
+
+	return b
+}
+func BlockAsBytes(block Block) ([]byte, error) {
+	var w bytes.Buffer
+	if _, err := xdr.Marshal(&w, &block); err != nil {
+		return nil, fmt.Errorf("error marshalling block ids %v", err)
+	}
+	return w.Bytes(), nil
+}
+
+func (t *BlockBuilder) acceptBlockData(){
+	for{
+		select {
+		case <-t.stopChan:
+			t.started = false
+			return
+
+		case id := <-t.beginRoundEvent:
+			txList := t.transactionQueue[:common.Min(len(t.transactionQueue), MaxTransactionsPerBlock)]
+			t.transactionQueue = t.transactionQueue[common.Min(len(t.transactionQueue), MaxTransactionsPerBlock):]
```

Comment:
seems like no coinbase tx is added here. also nothing mentioned about collecting fees

Reply:
no mining fees nor transaction fees are included in this milestone and this PR as well